src/OpenTelemetry/Trace/TracerProviderSdk.cs

```diff
@@ -204,6 +191,23 @@ protected override void Dispose(bool disposing)
             in ActivityCreationOptions<ActivityContext> options,
             Sampler sampler)
         {
+            if (Sdk.SuppressInstrumentation)
+            {
+                return ActivityDataRequest.None;
+            }
+
+            if (sampler is AlwaysOnSampler)

```

Comment:
I'm concerned about the perf here.
The original code will only check `AlwaysOnSampler` or `AlwaysOffSampler` once (during the ctor) and hard-wire the logic for all subsequent calls. With this change we will pay the check for every single activity - whether created or not.

Reply:
Oh, yes I see! I hadn't connected those dots. Will revert soon.

Reply:
I guess what I was going for here was to reduce some code duplication to some extent. I've reverted and made the `Sdk.SuppressInstrumentation` check in each spot. Alternatively, I could have done something like:

```
private static ActivityDataRequest ComputeActivityDataRequestIfNotSuppressInstrumentation(Func<ActivityDataRequest> computeActivityDataRequest)
{
    return !Sdk.SuppressInstrumentation ? computeActivityDataRequest() : ActivityDataRequest.None;
}
```

and used it like

```
if (sampler is AlwaysOnSampler)
{
    listener.GetRequestedDataUsingContext = (ref ActivityCreationOptions<ActivityContext> options) =>
        ComputeActivityDataRequestIfNotSuppressInstrumentation(() => ActivityDataRequest.AllDataAndRecorded);
}
```

meh?